### PR TITLE
Refactor DeviceManager and fix data-parallel workloads

### PR DESCRIPTION
**Pull request by umalesTT**
### Ticket
None.

### Problem description
Needed to refactor DeviceManager to be more usable - reasons were extensively discussed offline.

### What's changed
Refactored DeviceManager to resolve sharding strategy from mesh shape and mesh axis names. 
Used regex in DeviceManager to shard weights on level of model modules, making it usable for LLMs with many layers as well as for existing use case.
Merged forward and backward graph in mnist multi-chip workloads, circumventing [this issue](https://github.com/tenstorrent/tt-xla/issues/2375) where we previously had incorrectly replicated loss across model replicas, which made each replica use the same loss in backward pass.

### Checklist
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: blacksmith/tools/device_manager.py [hunk 1]
```diff
@@ -42,70 +36,91 @@ def _setup(self):
...
+        # Read mesh_shape from config.
+        mesh_shape = tuple(self.config.mesh_shape)
+
+        # Parse mesh_axis_names from config.
+        if isinstance(self.config.mesh_axis_names, (list, tuple)):
+            axis_names = tuple(self.config.mesh_axis_names)
         else:
-            supported_strategies = [f for f in ParallelStrategy]
-            raise ValueError(f"Invalid parallelism: {self.strategy}. Supported strategies: {supported_strategies}.")
+            axis_names = self.config.mesh_axis_names
```
[the remark this answers is not in the document]
Comment (mmilosevicTT) [reply blacksmith/tools/device_manager.py]: +1 on this

File: blacksmith/tools/device_manager.py [hunk 2]
```diff
@@ -42,70 +36,91 @@ def _setup(self):
...
 
     def _apply_tensor_parallelism(self, model: nn.Module) -> nn.Module:
+        """Apply tensor parallelism using regex pattern matching from config."""
         torch_xla.sync(wait=True)
 
-        sharding_specs = self.config.tp_sharding_specs or {}
-        for name, param in model.named_parameters():
-            if param.dim() == 0:
-                continue
+        # Get sharding patterns from config (list of [pattern, spec] pairs)
```
Comment (mmilosevicTT) [inline blacksmith/tools/device_manager.py]: super nit:
```suggestion
        # Get sharding patterns from config (list of [pattern, spec] pairs).
```

File: blacksmith/experiments/torch/mnist/configs.py [hunk 3]
```diff
@@ -64,11 +63,11 @@ class TrainingConfig(BaseModel):
...
-    parallelism_strategy: ParallelStrategy  # [single, data_parallel, tensor_parallel]
-    mesh_shape: str = Field(default="8,1")  # Used if parallelism_strategy != single
-    tp_sharding_specs: dict[str, tuple[Optional[str], Optional[str]]] = Field(
-        default_factory=dict
-    )  # Used for model tp sharding
+    mesh_shape: Optional[List[int]] = Field(default=None)  # e.g., [2, 1] for data parallel, [1, 2] for tensor parallel
+    mesh_axis_names: Optional[Union[Tuple[str, ...], List[str]]] = Field(default=None)  # e.g., ["data", "model"]
+
+    # Model sharding patterns (regex pattern based - matches module names)
+    model_sharding_patterns: Optional[List[List]] = Field(default=None)
```
[the remark this answers is not in the document]
Comment (umalesTT) [reply blacksmith/experiments/torch/mnist/configs.py]: Yes sure
